sdk/iot/core/inc/az_iot_provisioning_client.h

```diff
@@ -0,0 +1,310 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+/**
+ * @file az_iot_provisioning_client.h
+ *
+ * @brief definition for Azure Device Provisioning Service device-client SDK.
+ */
+
+#ifndef _az_IOT_PROVISIONING_CLIENT_H
+#define _az_IOT_PROVISIONING_CLIENT_H
+
+#include <az_result.h>
+#include <az_span.h>
+#include <stdbool.h>
+#include <stdint.h>
+
+#include <_az_cfg_prefix.h>
+
+/**
+ * @brief Azure IoT Provisioning Client options.
+ *
+ */
+typedef struct az_iot_provisioning_client_options
+{
+  az_span user_agent; /** @brief The user-agent is a formatted string that will be used for Azure
+                         IoT usage statistics. */
+} az_iot_provisioning_client_options;
+
+/**
+ * @brief Azure IoT Provisioning Client.
+ *
+ */
+typedef struct az_iot_provisioning_client
+{
+  struct
+  {
+    az_span global_endpoint_hostname;
+    az_span id_scope;
+    az_span registration_id;
+    az_iot_provisioning_client_options options;
+  } _internal;
+} az_iot_provisioning_client;
+
+/**
+ * @brief Gets the default Azure IoT Provisioning Client options.
+ * @details Call this to obtain an initialized #az_iot_provisioning_client_options structure that
+ *          can be afterwards modified and passed to #az_iot_provisioning_client_init.
+ *
+ * @return #az_iot_provisioning_client_options.
+ */
+az_iot_provisioning_client_options az_iot_provisioning_client_options_default();
+
+/**
+ * @brief Initializes an Azure IoT Provisioning Client.
+ *
+ * @param client
+ * @param global_endpoint_hostname The Global device endpoint.
+ * @param id_scope The ID Scope.
+ * @param registration_id The Registration ID. This must match the client certificate name (CN part
+ *                        of the certificate subject).
+ * @param options A reference to an #az_iot_provisioning_client_options structure. Can be NULL.
+ * @return #az_result
+ */
+AZ_NODISCARD az_result az_iot_provisioning_client_init(
+    az_iot_provisioning_client const* client,

```

Comment:
>az_iot_provisioning_client const* client [](start = 4, length = 40)

Here you are initializing the client (write operation), but type is pointer to const  client ?

Reply:
yes the pointer asterisk should be on the struct not const:
[C Gibberish](https://www.cdecl.org/?q=int+const*+az_client)

Reply:
Good catch: replace all issue. I'll remove the pointer to const.

@danewalton I've interpreted what you said as having a const pointer to a non-const struct instead of the current (obviously wrong) pointer to a const struct. (Please don't write C in English ever again :)) ).  It sounds like a good idea to require the pointers to be const but it should be further discussed as it would require changes everywhere (i.e. why not have const pointers to const data?). You may want to open another issue.